app/models/manageiq/providers/openstack/network_manager/network_port.rb

```diff
@@ -1,2 +1,12 @@
 class ManageIQ::Providers::Openstack::NetworkManager::NetworkPort < ::NetworkPort
+
+  def disconnect_port
+    # Some ports link subnets to routers, so
+    # sever that association if the port is removed
+    cloud_subnets.each do |subnet|
```

Comment:
These might be the wrong subnets? I think we fill these in only for Vm's network ports?

We should look at ```belongs_to :device, :polymorphic => true``` , that relations for subnets is being fixed in https://github.com/ManageIQ/manageiq-providers-openstack/pull/188/files (could you verify it actually returns subnet here?)

Also we should check if we can conventionalize this by subnet.network_router = nil, only if the router is the router mentioned in the NetowrkPort (we might not store this). To avoid nullifying when the subnet was just connected somewhere else.

Hm, this approach is getting more complex than it should.

Reply:
maybe we should even remodel the relation, to not use ```subnet.network_router```, but to just use the network_port and sending the query through it

Reply:
The router gets returned out of `find_device_object` correctly there. Both `parser.network_ports` and `find_device_object` use the `network_port.fixed_ips` to associate subnets so I was thinking those were the correct subnets. The tests pass correctly, at least. (though perhaps this catches unrelated subnets? I don't have a good mental model of all the circumstances in which subnets get attached to ports and routers.)

Reply:
Right so there is NetworkPort which is interface of a VM (Vm is the device) and there is a NetworkPort that is interface of router. I think the network_port.cloud_subnets works only for Vm's interfaces.

Reply:
It does appear to work here, so that must not be the case that it only works on VMs.

Reply:
hm, the spec looks correct. So it must work :-)

Reply:
Though a spec case where this might break is that the targeted refresh contains both removing of old interface and adding new one. That might cause CloudSubnet.first.network_router_id to be nil, while it should point to a new router. Could you add this spec scenario?

Reply:
Good catch, I'll write another spec.

Reply:
Wrote a spec for the situation where the port is removed and a new one is created in the same refresh. `router_id` becomes nil as you suspected, but I don't understand why exactly. 

I inspected what was happening when the custom delete is called, and it looks like at that time the replacement NetworkPort doesn't exist yet. So if the old NetworkPort is the only one in the database at the time, wouldn't the new NetworkPort have to be created afterwards and set the id again?